llvm/test/CodeGen/AArch64/vector-extract-last-active.ll

```diff
@@ -174,12 +174,12 @@ define i64 @extract_last_i64(<2 x i64> %data, <2 x i64> %mask, i64 %passthru) {
 ; SVE-FIXED-NEXT:    sub sp, sp, #16
 ; SVE-FIXED-NEXT:    .cfi_def_cfa_offset 16
 ; SVE-FIXED-NEXT:    cmtst v1.2d, v1.2d, v1.2d
-; SVE-FIXED-NEXT:    index z2.s, #0, #1
+; SVE-FIXED-NEXT:    index z3.s, #0, #1
 ; SVE-FIXED-NEXT:    mov x9, sp
 ; SVE-FIXED-NEXT:    str q0, [sp]
-; SVE-FIXED-NEXT:    xtn v1.2s, v1.2d
-; SVE-FIXED-NEXT:    and v2.8b, v1.8b, v2.8b
-; SVE-FIXED-NEXT:    umaxp v1.2s, v1.2s, v1.2s
+; SVE-FIXED-NEXT:    xtn v2.2s, v1.2d
+; SVE-FIXED-NEXT:    umaxv s1, v1.4s
```

Comment:
What's going on here?  Going from umaxp to umaxv doesn't seem like an improvement.

Reply:
In the originally generated code, both the `and` and `umaxp` instructions used the result of the `xtn` above. With my patch the truncation is elided for the `reduce_or` part, and the result of the `cmptst` is used directly by the `umaxv`. Yes, going from `umaxp` to `umaxv` is not ideal here, but we're removing the dependency on the result of `xtn`, which is an improvement. I'm not sure how this might affect performance in this case, but seeing that the result of this `umaxv` is used 6 instructions after this, and we don't have to wait for the `xtn` anymore I wouldn't be surprised if there was no difference. I could try to benchmark it to see if there's any difference.

Also, this codegen is specific to the `vector.extract.last.active` intrinsic (with 64-bit wide elements), which seems to be being actively worked on in https://github.com/llvm/llvm-project/pull/112738 and https://github.com/llvm/llvm-project/pull/118810.

Reply:
Oh, I see, we got lucky with instruction reuse before... and if we don't get the reuse, S-form umaxv is faster than xtn+umaxp?  In that case, I guess this change is okay.